pkg/sql/opt/exec/execbuilder/tests/local/local_test: TestExecBuild_sql_statistics_persisted failed

pkg/sql/opt/exec/execbuilder/tests/local/local_test.TestExecBuild_sql_statistics_persisted [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9171084?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9171084?buildTab=artifacts#/) on master @ [5f5cf265a8717da8ee327f41c0d06836d11ac8ec](https://github.com/cockroachdb/cockroach/commits/5f5cf265a8717da8ee327f41c0d06836d11ac8ec):


```
=== RUN   TestExecBuild_sql_statistics_persisted
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9986f69167dd2633633c749cd542ec4c/logTestExecBuild_sql_statistics_persisted2954968955
    test_log_scope.go:79: use -show-logs to present logs inline
    logic.go:2874: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11604/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/sql_statistics_persisted:73: EXPLAIN (VERBOSE)
        SELECT
          aggregated_ts,
          fingerprint_id,
          app_name,
          node_id,
          agg_interval,
          metadata,
          statistics
        FROM
          system.transaction_statistics
        expected:
              distribution: local
              vectorized: true
            ·
              • scan
                columns: (aggregated_ts, fingerprint_id, app_name, node_id, agg_interval, metadata, statistics)
                estimated row count: 1,000 (missing stats)
                table: transaction_statistics@primary
                spans: FULL SCAN
            
        but found (query options: "") :
            distribution: local
            vectorized: true
            ·
            • scan
              columns: (aggregated_ts, fingerprint_id, app_name, node_id, agg_interval, metadata, statistics)
              estimated row count: 1 (100% of the table; stats collected <hidden> ago)
              table: transaction_statistics@primary
              spans: FULL SCAN
[13:34:23] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11604/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/sql_statistics_persisted with config local: 3 tests, 1 failures
    logic.go:3972: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11604/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/sql_statistics_persisted:95: error while processing
    logic.go:3972: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11604/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/sql_statistics_persisted:95: too many errors encountered, skipping the rest of the input
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9986f69167dd2633633c749cd542ec4c/logTestExecBuild_sql_statistics_persisted2954968955
--- FAIL: TestExecBuild_sql_statistics_persisted (2.43s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecBuild_sql_statistics_persisted.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25709

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -1452,6 +1452,14 @@ func (t *logicTest) newCluster(
 	stats.DefaultAsOfTime = 10 * time.Millisecond
 	stats.DefaultRefreshInterval = time.Millisecond
 
+	// Disable stats collection on system tables before the cluster is started,
+	// otherwise there is a race condition where stats may be collected before we
+	// can disable them with `SET CLUSTER SETTING`. We disable stats collection on
+	// system tables in order to have deterministic tests.
+	stats.AutomaticStatisticsOnSystemTables.Override(
+		context.Background(), &params.ServerArgs.TempStorageConfig.Settings.SV, false,
+	)
+
 	t.cluster = serverutils.StartNewTestCluster(t.rootT, cfg.NumNodes, params)
 	if cfg.UseFakeSpanResolver {
 		fakeResolver := physicalplanutils.FakeResolverForTestCluster(t.cluster)
@@ -1637,11 +1645,6 @@ func (t *logicTest) newCluster(
 		// See #37751 for details.
 		if _, err := conn.Exec(
 			"SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false",
-		); err != nil {
-			t.Fatal(err)
-		}
-		if _, err := conn.Exec(
-			"SET CLUSTER SETTING sql.stats.system_tables_autostats.enabled = false",
 		); err != nil {
 			t.Fatal(err)
 		}

--- a/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze
+++ b/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze
@@ -124,18 +124,14 @@ regions: <hidden>
 statement ok
 GRANT SELECT ON crdb_internal.tables TO root;
 
-# For some reason, even though we explicitly disable the automatic stats
-# collection for both user and system tables, rarely we still see that the stats
-# on system.privileges were collected, so we filter out a single line from the
-# output to make the test deterministic.
 query T
-SELECT * FROM [EXPLAIN (VERBOSE) SELECT * FROM system.privileges WHERE path = 'vtable/crdb_internal/tables']
-  WHERE info NOT LIKE '%estimated row count%';
+EXPLAIN (VERBOSE) SELECT * FROM system.privileges WHERE path = 'vtable/crdb_internal/tables'
 ----
 distribution: local
 vectorized: true
 ·
 • scan
   columns: (username, path, privileges, grant_options, user_id)
+  estimated row count: 10 (missing stats)
   table: privileges@privileges_path_user_id_key
   spans: /"vtable/crdb_internal/tables"-/"vtable/crdb_internal/tables"/PrefixEnd

